core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -56,7 +60,9 @@ class LogManager(val logDirs: Array[File],
                  scheduler: Scheduler,
                  val brokerState: BrokerState,
                  brokerTopicStats: BrokerTopicStats,
-                 time: Time) extends Logging {
+                 val zkUtils: ZkUtils,
```

Comment:
@junrao Thanks for the suggestion!

One concern with having this queue for disk failure events is that, in the event a log directory fails, every attempt to access any replica in that log directory will generate a disk failure event for the same log directory. A lot of events may be instantiated and put into this queue which wastes CPU and memory. I am also concerned that we can not bound the number of events generated for a single disk failure since we don't know when the OS will schedule that thread to read event from this queue.

Maybe we don't need this `DiskFailureChannel`. We can just schedule a thread in `ReplicaManager` to read from `LogManager.offlineLogDirs` and call `ReplicaManager.handleLogDirs(dir)` when there is new offline log dir.

I understand that the goal of this alternative approach is to make sure that the state in `ReplicaManager` (e.g. `ReplicaManager.allPartitions`) is consistent with the state in `LogManager` (e.g. `LogManager.logs`). And I agree that we can achieve this goal by having an extra thread in `ReplicaManager`. But I am not sure that the benefit of this alternative approach is worth the extra thread. Note that the disadvantages of the alternative approach is: 1) It requires an extra thread in `ReplicaManager` which Kafka's Java class a bit more complicated; and 2) The log failure event will be processed asynchronously which potentially delays the controller notification and leader election. And we can not bound the delay.

Can you help me understand why it is necessary to keep state in `ReplicaManager` and `LogManager` consistent all the time? The current approach in the patch guarantees eventual consistency, i.e. the `ReplicaManager.handleLogDirFailure()` will be called after controller notification is sent and the broker receives `LeaderAndIsrRequest` from controller. It seems that nothing will go wrong during the period of inconsistency. And the alternative approach makes them consistent by delaying the execution of `LogManager.handleLogDirFailure()`.

I will implement the alternative approach using an extra thread if the consistency is more important than the extra complexity and the potential delay in log failure handling.